aten/src/ATen/native/cuda/FusedAdamWKernel.cu

```diff
@@ -29,16 +39,40 @@ void _fused_adamw_kernel_cuda_(
     const c10::optional<at::Tensor>& grad_scale,
     const c10::optional<at::Tensor>& found_inf
 ) {
-  if (amsgrad) {
-    TORCH_CHECK(
-        at::native::check_fast_path_restrictions({params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs}),
-        "params, grads, exp_avgs, exp_avg_sqs, and max_exp_avg_sqs must have same dtype, device, and layout");
-    _fused_adamw_amsgrad_cuda_impl_(params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs, state_steps, lr, beta1, beta2, weight_decay, eps, maximize, grad_scale, found_inf);
-  } else {
-    TORCH_CHECK(
-        at::native::check_fast_path_restrictions({params, grads, exp_avgs, exp_avg_sqs}),
-        "params, grads, exp_avgs, and exp_avg_sqs must have same dtype, device, and layout");
-    _fused_adamw_cuda_impl_(params, grads, exp_avgs, exp_avg_sqs, state_steps, lr, beta1, beta2, weight_decay, eps, maximize, grad_scale, found_inf);
+  auto device_grad_scale_map = init_map(grad_scale);
+  auto device_found_inf_map = init_map(found_inf);
+  const auto nested_tensorlists = [&]() -> std::vector<TensorList> {
+    if (amsgrad) {
+      return {params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs, state_steps};
+    } else {
+      return {params, grads, exp_avgs, exp_avg_sqs, state_steps};
+    }
+  }();
+  OptionalDeviceGuard guard;
+  for (const auto & device_map : group_tensors_by_device_and_scalartype(nested_tensorlists, /* has_state_steps */ true)) {
```

Comment:
Will you be switching this to call the generic foreach util once that's ready? I think that would be ideal (vs having a copy in the fused_adam_utils) and I'd be curious if the perf results would be different.